tests/pytests/integration/netapi/test_ssh_client.py

```diff
@@ -0,0 +1,431 @@
+    return str(sshd_server.config_dir / "client_key")
+
+
+@pytest.fixture(scope="module")
+def webserver():
+    with Webserver(handler=SaveRequestsPostHandler) as server:
+        yield server
+
+
+@pytest.fixture(scope="module")
+def webserver_root(webserver):
+    return webserver.web_root
+
+
+@pytest.fixture(scope="module")
+def webserver_handler(webserver):
+    return webserver.handler
+
+
+@pytest.fixture(scope="module")
+def salt_auth_account_1(salt_auth_account_1_factory):
+    with salt_auth_account_1_factory as account:
+        yield account
+
+
+@pytest.fixture(scope="module")
+def salt_auto_account(salt_auto_account_factory):
+    with salt_auto_account_factory as account:
+        yield account
+
+
+def test_ssh(client, auth_creds, salt_ssh_roster_file, rosters_dir, ssh_priv_key):
+    low = {
+        "client": "ssh",
+        "tgt": "localhost",
+        "fun": "test.ping",
+        "ignore_host_keys": True,
+        "roster_file": str(salt_ssh_roster_file),
+        "rosters": [rosters_dir],
+        "ssh_priv": ssh_priv_key,
+        **auth_creds,
+    }
+
+    ret = client.run(low)
+
+    assert "localhost" in ret
+    assert "return" in ret["localhost"]
+    assert ret["localhost"]["return"] is True
+    assert ret["localhost"]["id"] == "localhost"
+    assert ret["localhost"]["fun"] == "test.ping"
+
+
+def test_ssh_unauthenticated(client):
+    low = {"client": "ssh", "tgt": "localhost", "fun": "test.ping"}
+    with pytest.raises(EauthAuthenticationError):
+        client.run(low)
+
+
+def test_ssh_unauthenticated_raw_shell_curl(client, webserver_root, webserver_handler):
+
+    fun = "-o ProxyCommand curl {}".format(webserver_root)
+    low = {"client": "ssh", "tgt": "localhost", "fun": fun, "raw_shell": True}
+
+    with pytest.raises(EauthAuthenticationError):
+        client.run(low)
+
+    assert webserver_handler.received_requests == []
+
+
+def test_ssh_unauthenticated_raw_shell_touch(client, tmp_path):
+
+    badfile = tmp_path / "badfile.txt"
+    fun = "-o ProxyCommand touch {}".format(badfile)
+    low = {"client": "ssh", "tgt": "localhost", "fun": fun, "raw_shell": True}
+
+    with pytest.raises(EauthAuthenticationError):
+        client.run(low)
+
+    assert badfile.exists() is False
+
+
+def test_ssh_authenticated_raw_shell_disabled(client, tmp_path):
+
+    badfile = tmp_path / "badfile.txt"
+    fun = "-o ProxyCommand touch {}".format(badfile)
+    low = {"client": "ssh", "tgt": "localhost", "fun": fun, "raw_shell": True}
+
+    with patch.dict(client.opts, {"netapi_allow_raw_shell": False}):
+        with pytest.raises(EauthAuthenticationError):
+            client.run(low)
+
+    assert badfile.exists() is False
+
+
+def test_shell_inject_ssh_priv(
+    client, salt_ssh_roster_file, rosters_dir, tmp_path, salt_auto_account
+):
+    """
+    Verify CVE-2020-16846 for ssh_priv variable
+    """
+    # ZDI-CAN-11143
+    path = tmp_path / "test-11143"
+    tgt = "www.zerodayinitiative.com"
```

Comment:
How about we use https://pytest-httpserver.readthedocs.io/en/latest/ here or something similar, so we are not depending on external resources.

Reply:
Yep, that might be a good approach going forward.

Reply:
Did you want to handle that in this PR or a follow up one? I'm fine either way.

Reply:
This one still needs work, i have to revisit this.
But probably on a separate PR, where we address all webservers we use ...